FEM-compatible Project Builder (aka Lab 2 aka /lab-fem)

shaunanoordin opened the pull request:
## PR Overview

This PR adds an FEM-compatible Project Builder to PFE.

Context: as we try to migrate fully from PFE to FEM, we need to encourage project owners to build projects that are compatible with FEM. To do so, we need to give them a _Project Builder that's compatible with FEM._

- ~~The FEM-compatible Project Builder is accessible from the `/fem-lab` route.~~
- Code is based on the original Project Builder in the adjacent `./lab` folder.

### Testing

Staging URL: https://pr-6037.pfe-preview.zooniverse.org/

~~Suggested Test URLs:~~
- ~~https://pr-6037.pfe-preview.zooniverse.org/lab-fem/~~
- ~~https://pr-6037.pfe-preview.zooniverse.org/lab-fem/1873/workflows/3520~~

### Status

WIP

Completed:
- Added stripped-down Workflow Editor from the earlier `lab` branch that Jim worked on, plus its styles
- ~~Added new `/lab-fem` route (we can change this to `/lab-2` later if we decide to)~~
- ~~Example: https://local.zooniverse.org:3735/lab-fem/1873/workflows/3520~~
- UPDATE: we no longer explicitly split the FEM Lab and the PFE Lab via an explicit URL. Instead, the same PFE-compatible OR FEM-compatible workflow editor page is accessible from https://local.zooniverse.org:3735/lab/1873/workflows/3520 . Whether or not you get the PFE-compatible or FEM-compatible version depends on the underlying project, and whether it's configured to use FEM.


TODO:
- Shortcuts Editor still needs to be added to match the `lab` branch.
- ~~Holy crap a lot of the workflows component has the `/lab` path baked into its links, this needs to be solved.~~

shaunanoordin commented:
## PR Update

❗ Big change: whether or not your get an _FEM-compatible_ or _PFE-compatible_ Project Builder page is now determined by the underlying project's configuration, NOT an explicit URL.

e.g. https://pr-6037.pfe-preview.zooniverse.org/lab/1873/workflows/3532 will give you the FEM-compatible workflow editor, IF project 1873 is configured to use FEM. Otherwise, you'll get the PFE-compatible workflow editor.

- `FEMLabRouter` component added to perform the splitting logic. This 'splitter' returns a different component, depending on the path AND whether or not the associated Zooniverse Project is FEM-compatible.
- A LOT of massive code changes have been simplified, woo hoo!
- At the moment, the _"should this project use FEM-compatible pages?"_ configuration flag is... supremely arbitrary. (As of writing, the check is just _"is this Project 1873?"_ ) I'll address this next. See [Slack convo](https://zooniverse.slack.com/archives/C02HVGW0JAY/p1635515619025100) for ongoing discussion on where the config flag should be placed.

shaunanoordin commented:
## PR Update

- The _"should this project use FEM-compatible pages?"_ check is now determined by the following:
  - Is the project's slug registered in the PFE code? 
  - Is the `?femLab=true` flag set? e.g. https://local.zooniverse.org:3735/lab/1873/workflows/3532?femLab=true
- There is an alternate override, `?pfeLab=true`, to force using the PFE-compatible lab pages.

I'm going to mark this PR ready for review, but I want to have one additional quick chat with Cliff on Slack to see if he wants to have an easier toggle for enabling/disabling FEM lab mode (i.e. use Experimental Tools while we wait for the full top-level `projectResource.useFEM` flag to be set up.

shaunanoordin commented:
## PR Update (Ready-For-Review Compilation)

This PR adds an FEM-compatible Project Builder to PFE.

- New behaviour: the Project Builder will _automatically_ switch between FEM-compatible or PFE-compatible pages, _depending on the project_ being edited.
  - e.g. https://pr-6037.pfe-preview.zooniverse.org/lab/1873/workflows/3532 will give you the FEM-compatible workflow editor, IF project 1873 is configured to use FEM. Otherwise, you'll get the PFE-compatible workflow editor.
- Currently, the switching logic is as follows:
  - use FEM Lab if the project slug is registered as a monorepo route in the PFE code, OR if `?femLab=true` flag is set
  - use PFE Lab in all other cases, OR if `?pfeLab=true` flag is set

Context: as we try to migrate fully from PFE to FEM, we need to encourage project owners to build projects that are compatible with FEM. To do so, we need to give them a _Project Builder that's compatible with FEM._

### Dev Notes

- Switching logic is found in `FEMLabRouter.jsx`
- `app/pages/fem-lab/workflow.cjsx` is just a modified version of `app/pages/lab/workflow.cjsx`

### Status

This PR is now ready for review

shaunanoordin commented:
@lcjohnso @mrniaboc @snblickhan 

## Executive Summary

- With the current incarnation of the PR, the PFE Project Builder will _automagically_ switch between PFE-compatible pages and FEM-compatible pages.
- The project builder at `zooniverse.org/lab/1234/workflows/5678` will use **FEM-compatible Lab pages** if ANY one or more of the following conditions are met:
  - Project is already using the FEM Classifier (e.g. Planet Hunters TESS)
  - Project has `experimental_tools.femLab = true` (this can be set via Admin pages)

We can also add `?pfeLab=true` or `?femLab=true` to the URL to _override_ the automagic switching. Please only use to debug problems during this early migration efforts; I plan to remove these flags eventually.

Long term, I DO believe there's agreement among the devs to use a project-level `project.plsUseFEM = true` attribute. (I think that's stalled for the moment because it's all tied up on how to also route the FEM Classifier?) The experimental tool solution in this PR is just something to get this PR out the door. 👌 

Here are the pros/cons of our current approach:

- ➕  We can easily let project owners _start building/editing_ FEM-compatible projects by toggling the **FEM Lab.**
- ✖️ Unfortunately, the project won't have access to the **FEM Classifier**, since that still requires devs to set up routes on PFE and FrontDoor, etc.
- ✖️ ~~Also, without the FEM Classifier, the project owners can't test their projects.~~ _EDIT: wait I just realised I can route _project owners_ to `https://frontend.preview.zooniverse.org/` for testing. Huh! Should have thought of that earlier. OK, that's my next PR._
- ➕ When we finally have the project-level attribute set up, we can very easily "transfer" `project.experimental_tools.femLab = true` to `project.plsUseFEM = true`
- ❗ Did I mention that **all existing FEM projects will now use the FEM compatible Lab** ? Do we need to give them a heads up?

shaunanoordin commented:
Blargh, the staging deploys are still borked since this morning, so it's difficult to test the code on https://pr-6037.pfe-preview.zooniverse.org/ until it's fixed.

shaunanoordin commented:
Updated branch to reflect changes to `workflow.cjsx` in #6042

lcjohnso commented:
@shaunanoordin -- Following up on our conversation today -- a few opinions:
1. Do not auto-opt-in projects to using `lab-fem` that are already redirecting to FEM.  We want to finalize the FEM project builder interface changes before enabling for _any_ projects.
2. Both the "View Project" and "Test This Workflow" buttons should point to FEM.  Specifically, the fe-project.zooniverse.org domain.

eatyourgreens commented:
Can we use `front end.preview.zooniverse.org` for links, so that JS, images etc are served via the CDN?

shaunanoordin commented:
@eatyourgreens I wanted to confirm something that I was discussing with Cliff earlier. Can you please help confirm my understanding that...?

- `fe-project.zooniverse.org` points to the `production-release` version of FEM
- `frontend.preview.zooniverse.org` points to the `master` version of FEM

If what I told Cliff is correct, then the idea to use `fe-project.zooniverse.org` is to ensure project owners don't accidentally get a preview of "merged but not yet deployed" patches.

shaunanoordin commented:
Also, following Cliff's comments:

## PR Update

- all existing FEM projects will **not automatically use** the FEM compatible Lab.
  - This means we need to remember enable, e.g. `planetHuntersTessProject.experimental_tools = ['femLab']`
  - Code-wise, I've removed the `usesMonorepo()` clause in FEMLabRouter
- Noted that the "View Project" and "Test this Workflow" should both point to the FEM preview URL.
  - This will be done in _a separate PR,_ as discussed.

I've updated the [Executive Summary](https://github.com/zooniverse/Panoptes-Front-End/pull/6037#issuecomment-957979389) with some notes.

eatyourgreens commented:
Oh good point. `fe-project` bypasses our proxying so doesn’t quite match the production setup either. Probably not a big deal, but you can’t go to Talk, for example.

mcbouslog commented on app/pages/lab-fem/_classifier/tasks/shortcut/editor.jsx:
```diff
@@ -0,0 +1,140 @@
+import PropTypes from 'prop-types';
```
I'll suggest deleting this file and delete `ShortcutEditor` from app/pages/lab-fem/workflow.cjsx if we know it's deprecated. Seems like a lot to add just to show as deprecated.

mcbouslog commented on app/pages/lab-fem/workflow.cjsx:
```diff
@@ -0,0 +1,888 @@
+          {if @state.selectedTaskKey? and @props.workflow.tasks[@state.selectedTaskKey]?
+            TaskEditorComponent = taskComponents[@props.workflow.tasks[@state.selectedTaskKey].type]?.Editor
+            <div>
+              {if 'shortcut' in @props.project.experimental_tools
+                <ShortcutEditor workflow={@props.workflow} task={@props.workflow.tasks[@state.selectedTaskKey]}>
+                  <TaskEditorComponent
+                    workflow={@props.workflow}
+                    task={@props.workflow.tasks[@state.selectedTaskKey]}
+                    taskPrefix="tasks.#{@state.selectedTaskKey}"
+                    project={@props.project}
+                    onChange={@handleTaskChange.bind this, @state.selectedTaskKey}
+                  />
+                </ShortcutEditor>
```
I think ok to delete here, and related new component file app/pages/lab-fem/_classifier/tasks/shortcut/editor.jsx

shaunanoordin replied:
That's a good point, Mark - the ShortcutEditor is locked behind an experimental_tool anyway, so it should be safe to delete altogether instead of keeping it around (in such a messy form) just to have the "deprecated" text on it.

mcbouslog reviewed (changes requested):
This looks great!! Confirmed status quo, and...
- PFE project => PFE project builder
- PFE project => toggle experimental FEM builder => FEM builder
- PFE project => query param to FEM builder => FEM builder
- FEM project => query param to PFE builder => PFE builder

Requesting changes relating to query param check and omitting ShortcutEditor, but open to pushback.

shaunanoordin commented on app/pages/lab-fem/fem-lab-router.jsx:
```diff
@@ -0,0 +1,29 @@
+  // solution to figuring out whether a project is FEM-compatible.
+  const thisProjectUsesFEM = (  // Use FEM-compatible pages if...
+    props?.project?.experimental_tools?.includes('femLab')  // ...the project has the femLab experimental tool
+    || !!props?.location?.query?.femLab // ...OR ?femLab=true query param is set
```
Can do, commit incoming!

shaunanoordin commented:
## PR Update

Thanks for the review, @mcbouslog ! This PR has been updated with your feedback:

- ShortcutEditor is now completely removed from FEM-Lab. Since the `shortcut` feature was hidden behind experimental_tools, it should be safe to remove completely, instead of having the messy workaround code just to disable + plonk a "Deprecated" label on top of the feature)
- `?pfeLab` and `?femLab` are now more strict about checking for the string `"true"`, instead of checking for _the presence of any string_

👍

mcbouslog reviewed (approved):
LGTM
